src/io/flutter/run/daemon/DaemonApi.java

```diff
@@ -0,0 +1,340 @@
+      // It's an event.
+      if (listener != null) {
+        DaemonEvent.dispatch(obj, listener);
+      } else {
+        LOG.info("ignored event from Flutter daemon: " + json);
+      }
+      return;
+    }
+
+    final int id;
+    try {
+      id = primId.getAsInt();
+    } catch (NumberFormatException e) {
+      LOG.error("Unable to parse response from Flutter daemon", e);
+      return;
+    }
+
+    final JsonElement error = obj.get("error");
+    if (error != null) {
+      LOG.warn("Flutter process returned an error: " + json);
+      final Command cmd = takePending(id);
+      if (cmd != null) {
+        cmd.completeExceptionally(new IOException("unexpected response: " + json));
+      }
+    }
+
+    final JsonElement result = obj.get("result");
+    complete(id, result);
+  }
+
+  /**
+   * Completes the pending command with the given id.
+   */
+  private void complete(int id, @Nullable JsonElement result) {
+    final Command cmd = takePending(id);
+    if (cmd == null) return;
+    cmd.complete(result);
+  }
+
+  private @Nullable Command takePending(int id) {
+    final Command cmd;
+    synchronized(pending) {
+      cmd = pending.remove(id);
+    }
+    if (cmd == null) {
+      LOG.warn("received a response for a request that wasn't sent: " + id);
+      return null;
+    }
+    return cmd;
+  }
+
+  private <T> CompletableFuture<T> send(String method, @Nullable Params<T> params) {
+    // Synchronize on nextId to ensure that we send one command at a time and they are numbered in the order they are sent.
+    synchronized (nextId) {
+      final int id = nextId.getAndIncrement();
+      final Command<T> command = new Command<>(method, params, id);
+      final String json = command.toString();
+      synchronized (pending) {
+        pending.put(id, command);
+      }
+      callback.accept(json);
+      return command.done;
+    }
+  }
+
+  private static void sendCommand(String json, ProcessHandler handler) {
+    final PrintWriter stdin = getStdin(handler);
+    if (stdin == null) {
+      LOG.warn("can't write command to Flutter process: " + json);
+      return;
+    }
+    stdin.write('[');
+    stdin.write(json);
+    stdin.write("]\n");
+    if (stdin.checkError()) {
+      LOG.warn("can't write command to Flutter process: " + json);
+    }
+  }
+
+  private static @Nullable PrintWriter getStdin(ProcessHandler processHandler) {
+    final OutputStream stdin = processHandler.getProcessInput();
+    if (stdin == null) return null;
+    return new PrintWriter(new OutputStreamWriter(stdin, Charsets.UTF_8));
+  }
+
+  @SuppressWarnings("unused")
+  static class Result {
+    private int code;
+    private String message;
+
+    boolean ok() {
+      return code == 0;
+    }
+
+    int getCode() {
+      return code;
+    }
+
+    String getMessage() {
+      return message;
+    }
+  }
+
+  /**
+   * A pending command to a Flutter process.
+   */
+  private static class Command<T> {
+    final String method;
```

Comment:
Not a big deal but if this is nullable, maybe annotate it so?  (I notice the next field is annotated.)